[LT] Sync LTC branch changes on torch/csrc/lazy/core

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69012

Summary: Some changes to torch/csrc/lazy/core were done on the
lazy_tensor_staging branch (https://github.com/pytorch/pytorch/pull/68427).
Merge those back into the trunk.

Differential Revision: [D32708696](https://our.internmc.facebook.com/intern/diff/D32708696)

desertfire committed:
[LT] Sync LTC branch changes on torch/csrc/lazy/core

Summary: Some changes to torch/csrc/lazy/core were done on the
lazy_tensor_staging branch (https://github.com/pytorch/pytorch/pull/68427).
Merge those back into the trunck.

[ghstack-poisoned]

wconstab commented on torch/csrc/lazy/ts_backend/ts_lowering_context.h:
```diff
@@ -74,16 +74,48 @@ class TORCH_API TSLoweringContext : public LoweringContext {
+        bool ok = lowering_->Lower(node);
+        CHECK(ok) << "Failed to lower: " << *node;
+      }
+      // At this point the outpout better be present, otherwise there is an issue
```
output

wconstab commented on torch/csrc/lazy/ts_backend/ts_node_lowering.h:
```diff
@@ -3,14 +3,14 @@
 #include <torch/csrc/api/include/torch/jit.h>
 #include <torch/csrc/lazy/backend/lowering_context.h>
 
-namespace torch_lazy_tensors {
```
oops, i landed it this way?

desertfire replied:
Yep

wconstab commented on torch/csrc/lazy/backend/backend_interface.h:
```diff
@@ -119,7 +119,7 @@ class TORCH_API BackendRegistrar {
 };
 
-inline const BackendImplInterface* TORCH_API getBackend() {
+static inline const BackendImplInterface* getBackend() {
```
we don't need TORCH_API macro here?

desertfire replied:
For a static inline function in a header file, we don't need it. I removed it according to a lint complain.

alanwaketan commented on torch/csrc/lazy/core/hash.h:
```diff
@@ -137,6 +181,11 @@ static inline hash_t Hash(const hash_t& value) {
   return value;
 }
 
+template <typename T>
```
Aha, I have a WIP to add this to the master. Feel free to do so and I probably just add the unit tests.

desertfire committed:
Update on "[LT] Sync LTC branch changes on torch/csrc/lazy/core"


Summary: Some changes to torch/csrc/lazy/core were done on the
lazy_tensor_staging branch (https://github.com/pytorch/pytorch/pull/68427).
Merge those back into the trunck.

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708696).
